### PR TITLE
Proposed fix for #19 - output CSV

### PR DESCRIPTION
**Pull request by ArtPoon**
- adds `--csvfile` option to write results to a user-specified CSV file
- added some docstrings and comments
- some PEP8 adjustments here and there

### CODE REVIEW

File: sc2rf.py [hunk 1]
```diff
@@ -32,7 +33,9 @@
     '': (29534, 99999), # probably nothing, but need something to mark the end of N
 }
 
+
 class Interval:
+    """ A closed interval of non-negative integers, e.g., 27 or 28-30 """
```
Comment (lenaschimmel) [inline sc2rf.py:38]: This class can also represent open intervals, as stated in the argument description: "An Interval can be a single number ("3"), a closed interval ("2-5" ) or an open one ("4-" or "-7"). The limits are inclusive. Only positive numbers are supported."

(other than this, your comments on my code are spot on)

Comment (ArtPoon) [reply sc2rf.py:38]: whoops. I'll update on my end, I think it should propagate to this PR

Comment (lenaschimmel) [reply sc2rf.py]: BTW, when I checked if my code _actually_ supports open intervals, I found a bug (which only affected the interval for the `-s` arg) which I just fixed.

**Comment by ktmeaton** [comment]
Thank you for this excellent implementation! I'm currently getting an error when using `--unique 1` (for debugging mode). This is on d29ca7c. The test data are all publicly available recombinants: [nextclade.aligned.fasta.gz](https://github.com/lenaschimmel/sc2rf/files/8419715/nextclade.aligned.fasta.gz)

```bash
python3 sc2rf.py nextclade.aligned.fasta --unique 1
```

```bash
Traceback (most recent call last):
  File "/Drives/W/Projects/covid-19/analysis/ncov-recombinant/sc2rf/sc2rf.py", line 1087, in <module>
    main()
  File "/Drives/W/Projects/covid-19/analysis/ncov-recombinant/sc2rf/sc2rf.py", line 211, in main
    show_matches([used_examples[i] for i in matching_example_indices], samples, writer=writer)
  File "/Drives/W/Projects/covid-19/analysis/ncov-recombinant/sc2rf/sc2rf.py", line 737, in show_matches
    regions.append((start_coord, last_coord, prev_definitive_match))
UnboundLocalError: local variable 'last_coord' referenced before assignment
```

**Comment by ArtPoon** [comment]
Thanks for catching this exception @ktmeaton - I've pushed a basic fix but these test data are exposing other problems I'll need to work through

**Comment by lenaschimmel** [comment]
Hi @Vjimenez-vasquez, I moved your comment to a new issue (#27), because it is not related to the feature we are discussing in this one.

I will reply there in a moment…

**Comment by ktmeaton** [comment]
Continuing with the example data ([nextclade.aligned.fasta.gz](https://github.com/lenaschimmel/sc2rf/files/8419715/nextclade.aligned.fasta.gz)) I am running with:

```bash
python3 sc2rf.py nextclade.aligned.fasta \
  --csvfile sc2rf.csv \
  --parents 1-4 \
  --breakpoints 1-4 \
  --unique 2 \
  --max-intermission-length 3 \
  --max-intermission-count 999 \
  --mutation-threshold 0.5

# --parents 1-4: Allows uncertainty about the parent
# --breakpoints 1-4: Allows flexibility regarding "long" intermissions
# --max-intermission-length 3: If there are 3 or less consecutive substitutions, don't treat this as a true breakpoint
# --max-intermission-count 999: Never allow intermissions to become a true breakpoint
# --mutation-threshold 0.5: Majority vote, 50% of strains in a clade must have a mutation for it to be diagnostic. (Helps with resolving parental uncertainty sometimes)
```

I was wondering if it would be possible to add 2 additional columns: `breakpoints_regions` and `intermissions_regions` like the following example. Having the regions of the breakpoints (in addition to those of each clade) would be great for reporting and plotting.

| strain                                  | clades                            | intermissions | breakpoints | clades_regions                                                                                          | breakpoints_regions | intermissions_regions   |
| --------------------------------------- | --------------------------------- | ------------- | ----------- | ------------------------------------------------------------------------------------------------------- | ------------------- | ----------------------- |
| England/ALDP-325CFB2/2022               | Delta/21J,Omicron/BA.1/21K        | 0             | 1           | 210:18163\|Delta/21J,19220:29742\|Omicron/BA.1/21K                                                      | 18164:19219         |                         |
| England/QEUH-312A66E/2022               | Delta/21J,Omicron/BA.1/21K        | 0             | 1           | 210:18163\|Delta/21J,19220:29742\|Omicron/BA.1/21K                                                      | 18164:19219         |                         |
| England/ALDP-31AE19E/2022               | Delta/21J,Omicron/BA.1/21K        | 0             | 1           | 210:18163\|Delta/21J,19220:29742\|Omicron/BA.1/21K                                                      | 18164:19219         |                         |
| England/ALDP-2F86D33/2021               | Delta/21J,Omicron/BA.1/21K        | 0             | 1           | 210:18163\|Delta/21J,19220:29742\|Omicron/BA.1/21K                                                      | 18164:19219         |                         |
| England/PLYM-33BB5CD/2022               | Delta/21J,Omicron/BA.1/21K        | 0             | 1           | 210:21618\|Delta/21J,21762:29742\|Omicron/BA.1/21K                                                      | 21617:21761         |                         |
| England/PLYM-332E079/2022               | Delta/21J,Omicron/BA.1/21K        | 0             | 1           | 210:21618\|Delta/21J,21762:29742\|Omicron/BA.1/21K                                                      | 21617:21761         |                         |
| England/PLYM-32FEF76/2022               | Delta/21J,Omicron/BA.1/21K        | 0             | 1           | 210:21618\|Delta/21J,21762:29742\|Omicron/BA.1/21K                                                      | 21617:21761         |                         |
| England/PLYM-336A651/2022               | Delta/21J,Omicron/BA.1/21K        | 0             | 1           | 210:21618\|Delta/21J,21762:29742\|Omicron/BA.1/21K                                                      | 21617:21761         |                         |
| England/PLYM-332E510/2022               | Delta/21J,Omicron/BA.1/21K        | 0             | 1           | 210:21618\|Delta/21J,21762:29742\|Omicron/BA.1/21K                                                      | 21617:21761         |                         |
| USA/PA-CDC-LC0474055/2022               | Delta/21J,Omicron/BA.1/21K        | 2             | 1           | 210:21762\|Delta/21J,21846:21846\|Omicron/BA.1/21K,21987:21987\|Delta/21J,22578:29742\|Omicron/BA.1/21K | 21761:22577         | 21846:21847,21987:21988 |
| USA/NJ-CDC-IBX952397337138/2022         | Delta/21J,Omicron/BA.1/21K        | 2             | 1           | 210:21762\|Delta/21J,21846:21846\|Omicron/BA.1/21K,21987:21987\|Delta/21J,22578:29742\|Omicron/BA.1/21K | 21761:22577         | 21846:21847,21987:21988 |
| USA/NJ-CDC-IBX640654818289/2022         | Delta/21J,Omicron/BA.1/21K        | 2             | 1           | 210:21762\|Delta/21J,21846:21846\|Omicron/BA.1/21K,21987:21987\|Delta/21J,22578:29742\|Omicron/BA.1/21K | 21761:22577         | 21846:21847,21987:21988 |
| USA/TN-CDC-ASC210559252/2021            | Delta/21J,Omicron/BA.1/21K        | 2             | 1           | 210:21762\|Delta/21J,21846:21846\|Omicron/BA.1/21K,21987:21987\|Delta/21J,22578:29742\|Omicron/BA.1/21K | 21761:22577         | 21846:21847,21987:21988 |
| USA/MA-CDC-STM-HZEBR92XC/2022           | Delta/21J,Omicron/BA.1/21K        | 2             | 1           | 210:21762\|Delta/21J,21846:21846\|Omicron/BA.1/21K,21987:21987\|Delta/21J,22578:29742\|Omicron/BA.1/21K | 21761:22577         | 21846:21847,21987:21988 |
| USA/PA-CDC-LC0474301/2022               | Delta/21J,Omicron/BA.1/21K        | 2             | 1           | 210:21762\|Delta/21J,21846:21846\|Omicron/BA.1/21K,21987:21987\|Delta/21J,22578:29742\|Omicron/BA.1/21K | 21761:22577         | 21846:21847,21987:21988 |
| USA/VT-CDCBI-CRSP_4GHG6RLGKSBSHCYP/2022 | Omicron/BA.1/21K,Omicron/BA.2/21L | 0             | 1           | 241:15240\|Omicron/BA.1/21K,15714:29510\|Omicron/BA.2/21L                                               | 15241:15713         |                         |
| USA/VT-CDCBI-CRSP_2HATRRNZ5U4IKSJA/2022 | Omicron/BA.1/21K,Omicron/BA.2/21L | 0             | 1           | 241:15240\|Omicron/BA.1/21K,15714:29510\|Omicron/BA.2/21L                                               | 15241:15713         |                         |
| USA/VT-CDCBI-CRSP_ZLZZSKWI6WSZQXJR/2022 | Omicron/BA.1/21K,Omicron/BA.2/21L | 0             | 1           | 241:15240\|Omicron/BA.1/21K,15714:29510\|Omicron/BA.2/21L                                               | 15241:15713         |                         |
| USA/VT-CDCBI-CRSP_3IGFKAUB6STYVNPK/2022 | Omicron/BA.1/21K,Omicron/BA.2/21L | 0             | 1           | 241:15240\|Omicron/BA.1/21K,15714:29510\|Omicron/BA.2/21L                                               | 15241:15713         |                         |
| USA/VT-CDCBI-CRSP_HOVIFDUIEJBVOGLK/2022 | Omicron/BA.1/21K,Omicron/BA.2/21L | 0             | 1           | 241:15240\|Omicron/BA.1/21K,15714:29510\|Omicron/BA.2/21L                                               | 15241:15713         |                         |
| USA/VT-CDCBI-CRSP_LEYQGGQYKC7CTXDR/2022 | Omicron/BA.1/21K,Omicron/BA.2/21L | 0             | 1           | 241:15240\|Omicron/BA.1/21K,15714:29510\|Omicron/BA.2/21L                                               | 15241:15713         |                         |

**Comment by ArtPoon** [comment]
Hi @ktmeaton - that's changing the original specification of the issue.  I have to focus on fixing the problems with the current CSV results for this PR, and I can't work on that until I've pushed code in our own project :-)

**Comment by lenaschimmel** [comment]
I had to take a short break from developing Sc2rf, but I'm back now!

This looks like it might be ready to be merged? But I can't, as long as it's marked as "work in progress".

By the way, could you run `sc2rf --update-readme`? It will add the documentation for your new paramerer `--csv-file` to the README automatically, which you can then commit.

**Comment by ArtPoon** [comment]
Yeah I had to take a break from this as well, to focus on other code projects.  I switched the PR to draft mode because I was getting some funny CSV output when running the `nextclade.aligned.fasta` file:
```console
art@Wernstrom sc2rf % python3 sc2rf.py nextclade.aligned.fasta --csvfile pr25.csv
Finding mutations in nextclade.aligned.fasta: 100%|█████| 30/30 [00:00<00:00, 154.21it/s]


Potential recombinants between ['Omicron / BA.1 / 21K', 'Omicron / BA.2 / 21L']:

coordinates                                 1111111111111222222222222222222222222222222222222222222222222222222
                                 223445899990000123455789011112222222222222333333333333344445566666677777888889
                               26780133334580144581427419067892566666778899000000245568914450502557823338238885
                               47933828942362944389041165516480777788781899144567002905432600867370558880718881
                               10027416344469879705804035582670834968563225308535235944804930400079892347111230

genes                           1a                 1b     S                                  3aEM   6   7b9N    
ref                            CTCACGCTGCACCCCGCACTCCCCACACCCGTGTCTCAGAGTGCAAGAATCACTCCGCATCCCCCACGCAGATCACGGGA

Omicron / BA.1 / 21K           T••GT••GA••••T••AG•CTT••G•••TT••ACTCT•••••AACGAGTCAGTGAATATATTT•TGGA•C•••TTTAAC•
Omicron / BA.2 / 21L           TGT•TAT••TGTTTTAA•T•T•TTGTGT••AGA•TCTGACTGAACG•GTC•GTGAAT•TA•TTTT•GATCCTCTTTAACC

USA/VT-CDCBI-CRSP_4GHG6RLGKSBS T••GT••GA••••T••AG•CTTTTGTGT••AGA•TCTGACTNNNNNNNNN•GTGAAT•TA•TTTT•GATCCTCTTTAACC 1 BP
USA/VT-CDCBI-CRSP_2HATRRNZ5U4I T••GT••GA••••T••AG•CTTTTGTGT••AGA•TCTGNCTNNNNNNNNN•GTGAAT•TA•TTTT•GATCCTCTTTAACC 1 BP
USA/VT-CDCBI-CRSP_ZLZZSKWI6WSZ T••GT••GA••••T••AG•CTTTTGTGT••AGA•TCTGACTNAACG•NNN•GTGAAT•TA•TTTT•GATCCTCTTTAACC 1 BP
USA/VT-CDCBI-CRSP_3IGFKAUB6STY T••GT••GA••••T••AG•CTTTTGTGT••AGA•TCTGACTGAACG•GTC•GTGAAT•TA•TTTT•GATCCTCTTTAACC 1 BP
USA/VT-CDCBI-CRSP_HOVIFDUIEJBV T••GT••GA••••T••AG•CTTTTGTGT••AGA•TCTGNNNNNNNNNNNN•GTGAAT•TA•TTTT•GATCCTCTTTAACC 1 BP
USA/VT-CDCBI-CRSP_LEYQGGQYKC7C T••GT••GA••••T••AG•CTTTTGTGT••AGA•TCTGACTNNNNNNNNN•GTGAAT•TA•TTTT•GATCCTCTTTAACC 1 BP

made with Sc2rf - available at https://github.com/lenaschimmel/sc2rf

art@Wernstrom sc2rf % head -n8 pr25.csv
sample,examples,intermissions,breakpoints,regions
USA/VT-CDCBI-CRSP_4GHG6RLGKSBSHCYP/2022,"Omicron/BA.1/21K,Omicron/BA.2/21L",0,1,"241:15240|Omicron/BA.1/21K,15714:29510|Omicron/BA.2/21L"
USA/VT-CDCBI-CRSP_2HATRRNZ5U4IKSJA/2022,"Omicron/BA.1/21K,Omicron/BA.2/21L",0,1,"241:15240|Omicron/BA.1/21K,15714:29510|Omicron/BA.2/21L"
USA/VT-CDCBI-CRSP_ZLZZSKWI6WSZQXJR/2022,"Omicron/BA.1/21K,Omicron/BA.2/21L",0,1,"241:15240|Omicron/BA.1/21K,15714:29510|Omicron/BA.2/21L"
USA/VT-CDCBI-CRSP_3IGFKAUB6STYVNPK/2022,"Omicron/BA.1/21K,Omicron/BA.2/21L",0,1,"241:15240|Omicron/BA.1/21K,15714:29510|Omicron/BA.2/21L"
USA/VT-CDCBI-CRSP_HOVIFDUIEJBVOGLK/2022,"Omicron/BA.1/21K,Omicron/BA.2/21L",0,1,"241:15240|Omicron/BA.1/21K,15714:29510|Omicron/BA.2/21L"
USA/VT-CDCBI-CRSP_LEYQGGQYKC7CTXDR/2022,"Omicron/BA.1/21K,Omicron/BA.2/21L",0,1,"241:15240|Omicron/BA.1/21K,15714:29510|Omicron/BA.2/21L"
USA/PA-CDC-LC0474055/2022,"Delta/21J,Omicron/BA.1/21K,Omicron/BA.2/21L",4,1,"210:241|Delta/21J,670:670|Delta/21J,2790:2790|Delta/21J,2832:4181|Delta/21J,4184:4184|Delta/21J,4321:4321|Delta/21J,5386:7124|Delta/21J,8393:9053|Delta/21J,9344:9344|Delta/21J,9424:9424|Delta/21J,9534:9534|Delta/21J,9866:10029|Delta/21J,10198:10198|Delta/21J,10447:11332|Delta/21J,11537:11537|Delta/21J,12880:12880|Delta/21J,13195:14408|Delta/21J,15240:15451|Delta/21J,15714:16466|Delta/21J,17410:19220|Delta/21J,19955:19955|Delta/21J,20055:21618|Delta/21J,21762:21762|Delta/21J,21846:21846|Omicron/BA.1/21K,21987:21987|Omicron/BA.2/21L,22200:22200|Omicron/BA.2/21L,22578:22578|Omicron/BA.2/21L,22673:22673|Omicron/BA.1/21K,22674:22674|Omicron/BA.1/21K,22679:22679|Omicron/BA.1/21K,22686:22686|Omicron/BA.1/21K,22688:22688|Omicron/BA.1/21K,22775:22775|Omicron/BA.1/21K,22786:22786|Omicron/BA.1/21K,22813:22882|Omicron/BA.2/21L,22917:22917|Omicron/BA.2/21L,22992:22995|Omicron/BA.2/21L,23013:23013|Omicron/BA.2/21L,23040:23040|Omicron/BA.2/21L,23048:23048|Omicron/BA.1/21K,23055:23055|Omicron/BA.1/21K,23063:23063|Omicron/BA.1/21K,23075:23403|Omicron/BA.1/21K,23525:23525|Omicron/BA.1/21K,23599:23599|Omicron/BA.1/21K,23604:23604|Omicron/BA.1/21K,23854:23854|Omicron/BA.1/21K,23948:24130|Omicron/BA.1/21K,24410:24410|Omicron/BA.1/21K,24424:24424|Omicron/BA.1/21K,24469:24503|Omicron/BA.1/21K,25000:25000|Omicron/BA.1/21K,25469:25469|Omicron/BA.1/21K,25584:25584|Omicron/BA.1/21K,26060:26060|Omicron/BA.1/21K,26270:26530|Omicron/BA.1/21K,26577:26577|Omicron/BA.1/21K,26709:26709|Omicron/BA.1/21K,26767:26767|Omicron/BA.1/21K,26858:26858|Omicron/BA.1/21K,27259:27259|Omicron/BA.1/21K,27382:27382|Omicron/BA.1/21K,27383:27383|Omicron/BA.1/21K,27384:27384|Omicron/BA.1/21K,27638:27638|Omicron/BA.1/21K,27752:27752|Omicron/BA.1/21K,27807:27807|Omicron/BA.1/21K,27874:27874|Omicron/BA.1/21K,28271:28271|Omicron/BA.1/21K,28311:28311|Omicron/BA.1/21K,28461:28461|Omicron/BA.1/21K,28881:28881|Omicron/BA.1/21K,28882:28882|Omicron/BA.1/21K,28883:28883|Omicron/BA.1/21K,28916:28916|Omicron/BA.1/21K,29402:29402|Omicron/BA.1/21K,29510:29510|Omicron/BA.1/21K,29742:29742|Omicron/BA.1/21K"
```

**Comment by ArtPoon** [comment]
Okay I found and resolved the problem.  I messed up trying to implement the user requested feature of reporting multiple matches in a region, *e.g.*, `20582:29132|21L&21M` - this breaks the requirement of definitive matches.  Ready for review, and sorry for the delay!

**Comment by ArtPoon** [comment]
@lenaschimmel - also I ran `--update-readme` as requested, looks like my PR is stale on a few of lines - sorry about that.

**Comment by lenaschimmel** [comment]
Thank you a lot, @ArtPoon, for your PR with this new feature! Also, the comments to my mostly uncommented code are highly appreciated. I merged it yesterday.

@ktmeaton Regarding your proposal for additional columns, would you like to create a separate issue for this? Not sure if my repo or Art's fork would be the right place for it, though.

**Comment by ArtPoon** [comment]
@lenaschimmel - very happy to help!

**Pull request merged**